### PR TITLE
ci: add ruff for print checks

### PR DESCRIPTION
**Pull request by diivi**
### Description
Recently committed a print statement, which reminded me of https://github.com/borgbase/vorta/pull/1612#issuecomment-1447782606

### Screenshots (if appropriate):
image

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [x] I have read the [CONTRIBUTING](https://vorta.borgbase.com/contributing/) guide.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

*I provide my contribution under the terms of the [license](./../../LICENSE.txt) of this repository and I affirm the [Developer Certificate of Origin][dco].*

[dco]: https://developercertificate.org/

### CODE REVIEW

File: setup.cfg
Comment (diivi) [inline setup.cfg:1]: line 42: `paramiko` should probably be removed here now?

File: pyproject.toml [hunk 1]
```diff
@@ -15,3 +15,6 @@ src_paths = ["src", "tests"]
 [build-system]
 requires = ["setuptools"]
 build-backend = "setuptools.build_meta"
+
+[tool.ruff]
+select = ["T"]
```
Comment (Hofer-Julian) [inline pyproject.toml:20]: In order to see if flake8 can replace flake8 and isort mid-term, let's activate all flake8 and isort features.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -24,7 +24,7 @@ jobs:
       run: |
         pip install .
         pip install -r requirements.d/dev.txt
-    - name: Test formatting with Flake8, isort and Black
+    - name: Test formatting with Flake8, isort, ruff and Black
       run: make lint
```
Comment (Hofer-Julian) [inline .github/workflows/test.yml:28]: I feel like pre-commit should replace the `make lint` rule.
This can also be handled in a follow up PR.

**Review by Hofer-Julian — COMMENTED** [review]
Thanks, looks good in general.

File: Makefile [hunk 3]
```diff
@@ -60,6 +60,7 @@ lint:
 	flake8
 	isort --check-only .
```
Comment (Hofer-Julian) [inline Makefile:61]: Even if we wouldn't want to immediately move away from flake8 as requested by @real-yfprojects, I think we can safely move from isort to ruff.

File: pyproject.toml (hunk 1 above)
Comment (real-yfprojects) [reply pyproject.toml:20]: *ruff

Comment (Hofer-Julian) [reply pyproject.toml:20]: Ah yes :P

File: package/fix_app_qt_folder_names_for_codesign.py
Comment (real-yfprojects) [inline package/fix_app_qt_folder_names_for_codesign.py:1]: Is it possible to limit the print checks to production code?

Comment (Hofer-Julian) [reply package/fix_app_qt_folder_names_for_codesign.py:1]: Yes https://beta.ruff.rs/docs/settings/#exclude

Comment (Hofer-Julian) [reply package/fix_app_qt_folder_names_for_codesign.py:1]: In that case, the status quo is the best option as far as I can see

Comment (diivi) [reply package/fix_app_qt_folder_names_for_codesign.py:1]: https://beta.ruff.rs/docs/settings/#per-file-ignores ?

Comment (m3nu) [reply package/fix_app_qt_folder_names_for_codesign.py:1]: Helper scripts don't need to be included in these checks. Just `src`.

File: pyproject.toml (hunk 1 above)
Comment (diivi) [reply pyproject.toml:20]: Turns out ruff does not use the same options with isort as isort itself. I could replicate the functionality to limit the changes to a minimum. But I can't figure out what option is triggering changes like these:
image
here are all the options ruff allows configuring - https://beta.ruff.rs/docs/settings/#isort
Also, I had to add a rule to remove line breaks between import sections too, why did we suppress them in the first place? They look better imo:
image

File: Makefile (hunk 3 above)
Comment (real-yfprojects) [reply Makefile:61]: Having https://github.com/borgbase/vorta/pull/1666#discussion_r1143728743 in mind I suggest keeping this as a check.

File: pyproject.toml (hunk 1 above)
Comment (real-yfprojects) [reply pyproject.toml:20]: I like them as well. But I guess @m3nu doesn't.

Comment (real-yfprojects) [reply pyproject.toml:20]: > But I can't figure out what option is triggering changes like these

Maybe `PyQt5` isn't recognized as a [third party library](https://beta.ruff.rs/docs/settings/#known-third-party)?

Comment (m3nu) [reply pyproject.toml:20]: Either way is good. If there are many imports, some line breaks can separate the various groups of imports.

**Comment by m3nu** [comment]
What's missing to move this forward?

**Comment by diivi** [comment]
> What's missing to move this forward?

Just a few minor changes:
* keeping isort and not using ruff's isort.
* include only src in checks

File: pyproject.toml (hunk 1 above)
Comment (diivi) [reply pyproject.toml]: I removed isort from the codebase and replaced it with ruff. It added some lines between imports but I think this separation is better.

**Comment by real-yfprojects** [comment]
It would be great if you could seperate your changes from the autogenerated ones. This allow excluding the latter from git blame.

**Comment by real-yfprojects** [comment]
> You mean make all config changes, skip pre-commit and push directly?

Maybe run pre-commit on the config changes before committing those.

> How will the sorting changes be made then?

Are you talking about the changes by *rust*?

> shouldn't git blame just blame me for the empty lines, and the other stuff remains the same?

I didn't look into the exact changes prior to writing my comment. This is just the general process of changing autoformatter configs. However rust didn't only add new lines I think.

**Comment by diivi** [comment]
> rust
ruff 😅 

So you mean I should make 2 commits right - one with my changes and the other one with ruff's changes?

**Comment by diivi** [comment]
Config changes - https://github.com/borgbase/vorta/pull/1666/commits/5d8073b8947b85d9c1ed3ce3a6d707c99893b39c.
I'll handle the merge conflicts after this is checked.

**Comment by real-yfprojects** [comment]
> I'll handle the merge conflicts after this is checked.

Looking good. After resolving merge conflicts (I think rebasing makes the most sense) I would squash the first two commits. Then a commit can be created that adds `run formatter` to `.git-blame-ignore-revs`.

**Comment by diivi** [comment]
Tried doing that, but I think I messed something up here - https://github.com/borgbase/vorta/compare/master...diivi:test-ruff-commits?expand=1.
I'll try to redo the changes in order.

**Comment by real-yfprojects** [comment]
> Tried doing that, but I think I messed something up

1. Make sure you local master is up-to-date with `upstream/master`.
2. Checkout this branch
3. `git rebase -i master`
4. In the file opened select to squash `ca90d7d373e4dfdd14ce2b10017516937bf43c9c` and to drop `a5d433fb252ddb8b0c97b803fc4eba58dd3e161d`
5. Save and close the file
6. Rerun the automatic formatters and commit the changes to the code base
7. Add the hash of the last commit to `.git-blame-ignore-revs` and commit it.

**Comment by real-yfprojects** [comment]
Merging this without the commit containing the formatting changes will break ci afaik.

**Comment by m3nu** [comment]
Wasn't this to find print statements when it started? 😀

And wasn't Black doing the formatting? Can't we just keep the same settings?

Anyways, if everything needs to be reformatted again, I suggest to just add your config changes and apply the formatting later when there aren't many pending PRs. Better to rerun the formatter than deal with all the conflicts.

**Comment by diivi** [comment]
> Wasn't this to find print statements when it started? grinning

Yeah, now it's about replacing isort with ruff too 😅 https://github.com/borgbase/vorta/pull/1666#discussion_r1143116959

**Comment by m3nu** [comment]
Understood. Then I suggest this:

- Please all help testing #1685 (Qt6 upgrade) so we can merge it today or tomorrow. This will start a new major version, since a major dependency changes.
- For this PR, please remove the formatting commits and only keep the commit that changes the tooling config.
- After Qt6 is merged, you can add the formatting changes without worrying about conflicts.
- Since this is a new major version, one could consider expanding the use of Ruff and replacing some more tools, if you like.

**Comment by m3nu** [comment]
#1685 is merged. So good to proceed here and add the formatting changes needed.

**Comment by real-yfprojects** [comment]
Does everyone approve this PR? Then I'll take care of merging this correctly.

**Comment by m3nu** [comment]
Good to go! Formatting change separate, if possible. 🙏

**Comment by real-yfprojects** [comment]
> Formatting change separate

I am planning on fast forwarding master with 3 commits corresponding to the commits of this PR.

File: Makefile [hunk 4]
```diff
@@ -58,8 +58,8 @@ flatpak-install: translations-to-qm
 
 lint:
 	flake8
-	isort --check-only .
 	black --check .
+	ruff check . --fix
```
Comment (real-yfprojects) [inline Makefile:62]: Why `--fix`? This is intended as a check only. That's why black is run with `--check`.

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -62,6 +57,11 @@ repos:
     hooks:
       - id: flake8
 
+  - repo: https://github.com/charliermarsh/ruff-pre-commit
+    rev: 'v0.0.257'
+    hooks:
```
Comment (real-yfprojects) [inline .pre-commit-config.yaml:62]: I'd prefer running the pre-commit hook with `--fix`.

**Pull request merged**